### PR TITLE
8233697: CHT: Iteration parallelization

### PR DESCRIPTION
**Pull request by walulyai**
Hi,

Please review this change to add parallel iteration of the ConcurrentHashTable. The iteration should be done during a safepoint without concurrent modifications to the ConcurrentHashTable.

Usecase is in parallelizing the merging of large remsets for G1.

Some background: The problem is that particularly during (G1) mixed gc it happens that the distribution of contents in the CHT is very unbalanced - young gen regions have a very small remembered set (little work), and old gen regions very large ones (much work).

Since the current work distribution is based on whole remembered sets (i.e. CHTs), this makes for a very unbalanced merge remsets phase in G1 when you have quite a bit more than the number of old gen regions threads at your disposal.
This negatively impacts pause time predictions (and obviously pause times are longer than necessary as many threads are idling to wait for the phase to complete).

This change only adds the infrastructure code in the CHT, there will be a follow-up with G1 changes.

Testing: tier 1-3

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8233697](https://bugs.openjdk.org/browse/JDK-8233697): CHT: Iteration parallelization

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10759/head:pull/10759` \
`$ git checkout pull/10759`

Update a local copy of the PR: \
`$ git checkout pull/10759` \
`$ git pull https://git.openjdk.org/jdk pull/10759/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10759`

View PR using the GUI difftool: \
`$ git pr show -t 10759`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10759.diff">https://git.openjdk.org/jdk/pull/10759.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp [hunk 1]
```diff
@@ -1298,4 +1339,75 @@ inline bool ConcurrentHashTable<CONFIG, F>::
...
+  size_t _claim_size_log2; // Log number of buckets in claimed range.
+  size_t _limit;      // Limit to number of claims
+
+  // If there is a paused resize, we also need to operate on the already resized items.
+  volatile size_t _next_to_claim_new_table;
+  size_t _claim_size_log2_new_table;
+  size_t _limit_new_table;
+
+public:
+  BucketsClaimer(ConcurrentHashTable<CONFIG, F>* cht) :
```
Comment (tschatzl) [inline src/hotspot/share/utilities/concurrentHashTable.inline.hpp:1359]: Would it be possible to add a claim size parameter (with default value `DEFAULT_CLAIM_SIZE_LOG2`)?

Particularly I'm not convinced that the default value is good :)

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp [hunk 2]
```diff
@@ -1298,4 +1339,75 @@ inline bool ConcurrentHashTable<CONFIG, F>::
...
 
+template <typename CONFIG, MEMFLAGS F>
+class ConcurrentHashTable<CONFIG, F>::BucketsClaimer {
+  // Default size of _claim_size_log2
+  static const size_t DEFAULT_CLAIM_SIZE_LOG2 = 7;
+
+  ConcurrentHashTable<CONFIG, F>* _cht;
+  // The table is split into ranges, every increment is one range.
+  volatile size_t _next_to_claim;
+  size_t _claim_size_log2; // Log number of buckets in claimed range.
```
Comment (tschatzl) [inline src/hotspot/share/utilities/concurrentHashTable.inline.hpp:1350]: For naming the constants containing log values (if kept), I would prefer if the existing style to put the `log2` in the front were kept.

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp [hunk 3]
```diff
@@ -1298,4 +1339,75 @@ inline bool ConcurrentHashTable<CONFIG, F>::
...
+    _next_to_claim(0),
+    _claim_size_log2(DEFAULT_CLAIM_SIZE_LOG2),
+    _limit(0),
+    _next_to_claim_new_table(0),
+    _claim_size_log2_new_table(0),
+    _limit_new_table(0)
+  {
+    size_t size_log2 = _cht->_table->_log2_size;
+    _claim_size_log2 = MIN2(_claim_size_log2, size_log2);
+    _limit = (size_t)1 << (size_log2 - _claim_size_log2);
```
Comment (tschatzl) [inline src/hotspot/share/utilities/concurrentHashTable.inline.hpp:1370]: Is it really advantageous to keep these values as log2? It seems to me that just storing the actual non-log values would actually require less work everywhere.
I am fine with restricting claim increments to powers of two (but even that seems artificial), there only seems to be no advantage to use logs anywhere (only that every use needs to do the shift to get the actual value).

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp [hunk 4]
```diff
@@ -1298,4 +1298,118 @@ inline bool ConcurrentHashTable<CONFIG, F>::
...
+      size_t claimed = Atomic::fetch_and_add(&claimer->_next, claimer->_size);
+      if (claimed < claimer->_limit) {
+        *start = claimed;
+        *stop  = MIN2(claimed + claimer->_size, claimer->_limit);
+        *table = claimer->_table;
+        return true;
+      }
+    }
+    return false;
+  }
```
Comment (tschatzl) [inline src/hotspot/share/utilities/concurrentHashTable.inline.hpp:1358]: Minor nit: A better place for this method is probably `InternalTableClaimer`; then you can also probably improve encapsulation (visibility) of the members if you want, but I'm good with keeping everything public (as default for a struct) for this helper class.
I did not think through the suggestion earlier.

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp (hunk 4 above)
Comment (walulyai) [reply src/hotspot/share/utilities/concurrentHashTable.inline.hpp]: you are right, i have updated, not certain about the `_table_claimer` and `_new_table_claimer` names, but couldn't come up with better

**Review by tschatzl — APPROVED** [review]

**Comment by robehn** [comment]
Hey!

Have you looked in:
concurrentHashTableTasks.inline.hpp
It contains BucketsOperation which is a base to do segmented work (very similar to BucketsClaimer).
The segmented work may be done with multiple threads.

**Comment by walulyai** [comment]
> Hey!
> 
> Have you looked in: concurrentHashTableTasks.inline.hpp It contains BucketsOperation which is a base to do segmented work (very similar to BucketsClaimer). The segmented work may be done with multiple threads.

I missed this, thanks for pointing it out. We should be able to extend the same for the scan operation.

**Review by robehn — APPROVED** [review]
Looks good, thanks!

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Comment by walulyai** [comment]
Thanks @robehn and @tschatzl for the reviews!

/integrate

**Pull request closed without merging**